File: Assets/FontWithPosition.ts
/** @noSelfInFile */

import { AlignMode, Font, Image } from "love.graphics";
import { isEmpty } from "Helpers";
import IconAsset from "./IconAsset";

export enum Format {
  LEFT,
  CENTER,
  RIGHT,
}

export enum OutlineThickness {
  NONE = 0,
  THIN = 1,
  THICK = 2,
}

export enum Fonts {
  STANDARD = "Assets/Fonts/Germania.ttf",
  FANTASY = "Assets/Fonts/dpcomic.ttf",
  ELOQUENT = "Assets/Fonts/Bitmgothic.ttf",
}

interface ConstructionOptions {
  font?: Fonts;
  size?: number; // Powers of 9 look really crisp
  format?: Format;
  icon?: IconAsset;
  iconFormat?: Omit<Format, Format.CENTER>;
  isDisabled?: boolean;
  outlineThickness?: OutlineThickness;
  color?: [number, number, number, number];
  limit?: number;
  alignMode?: AlignMode;
}

export default class FontWithPosition {
  id: string;
  x: number;
  y: number;
  text: string;
  format: Format;
  font: Font;
  icon?: IconAsset;
  iconFormat: Omit<Format, Format.CENTER> = Format.LEFT;
  isDisabled: boolean = false;
  color: [number, number, number, number] = [1, 1, 1, 1];
  outlineThickness: OutlineThickness;
  limit?: number;
  alignMode?: AlignMode;

  constructor(
    id: string,
    x: number,
    y: number,
    text: string,
    options?: ConstructionOptions
  ) {
    const size = options?.size ?? 9;
    const font = options?.font ?? Fonts.STANDARD;
    this.font = love.graphics.newFont(font, size);

    this.id = id;
    this.x = x;
    this.y = y;
    this.text = text;

    this.format = options?.format ?? Format.LEFT;
    this.icon = options?.icon;
    this.iconFormat =
      options?.iconFormat ??
      ((this.format === Format.CENTER ? Format.LEFT : this.format) as Omit<
        Format,
        Format.CENTER
      >);
    this.isDisabled = options?.isDisabled ?? false;
    this.outlineThickness = options?.outlineThickness ?? OutlineThickness.THIN;
    if (this.isDisabled) {
      this.setDisabled(true);
    }
    this.color = options?.color ?? [1, 1, 1, 1];
    this.limit = options?.limit;
    this.alignMode = options?.alignMode;
  }

  setDisabled(disabled: boolean): void {
    this.isDisabled = disabled;
    this.color = disabled ? [0.5, 0.5, 0.5, 1] : [1, 1, 1, 1];
  }

  printText(): void {
    love.graphics.setFont(this.font);

    const textW = this.font.getWidth(this.text);
    const textH = this.font.getHeight();
    const baseX = Math.floor(this.x - this.getFormatOffset(textW));
    const baseY = Math.floor(this.y - textH / 2);

    this.printOutline(baseX, baseY);

    love.graphics.setColor(this.color);
    !isEmpty(this.limit)
      ? love.graphics.printf(
          this.text,
          baseX,
          baseY,
          this.limit,
          this.alignMode ?? "left"
        )
      : love.graphics.print(this.text, baseX, baseY);
    this.renderIcon();
  }

  private printOutline(x: number, y: number): void {
    if (this.outlineThickness === OutlineThickness.NONE) {
      return;
    }

    love.graphics.setColor(0, 0, 0, this.color[3]);
    const offsets =
      this.outlineThickness === OutlineThickness.THICK
        ? [-2, -1, 0, 1, 2]
        : [-1, 0, 1];
    for (const ox of offsets) {
      for (const oy of offsets) {
        if (ox === 0 && oy === 0) {
          continue;
        }
        !isEmpty(this.limit)
          ? love.graphics.printf(
              this.text,
              x + ox,
              y + oy,
              this.limit,
              this.alignMode ?? "left"
            )
          : love.graphics.print(this.text, x + ox, y + oy);
      }
    }
  }

  private getFormatOffset(textW: number): number {
    switch (this.format) {
      case Format.LEFT:
        return 0;
      case Format.CENTER:
        return textW / 2;
      case Format.RIGHT:
        // If there's an icon that will be rendered to the right, account for its width
        const iconWidth =
          this.iconFormat === Format.RIGHT && !isEmpty(this.icon)
            ? this.icon.getWidth()
            : 0;
        return textW + iconWidth;
      default:
        return 0;
    }
  }

  private renderIcon(): void {
    if (isEmpty(this.icon)) {
      return;
    }
    love.graphics.setColor(this.color);

    // TODO: handle other align modes and formats
    switch (this.iconFormat) {
      case Format.LEFT:
        if (this.alignMode === "center") {
          love.graphics.draw(
            this.icon.image,
            this.x +
              this.limit! / 2 -
              this.font.getWidth(this.text) / 2 -
              this.icon!.getWidth() -
              2,
            this.y - this.icon.getHeight() / 2
          );
        } else {
          love.graphics.draw(
            this.icon.image,
            this.x - this.icon.getWidth() - 2,
            this.y - this.icon.getHeight() / 2
          );
        }
        break;
      case Format.RIGHT:
        love.graphics.draw(
          this.icon.image,
          this.x - this.icon.getWidth() + 1,
          this.y - this.icon.getHeight() / 2
        );
        break;
    }
  }
}

File: Assets/FontWithPosition.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local ____exports = {}
local ____TextManager = require("Assets.TextManager")
local TextManager = ____TextManager.default
local ____Helpers = require("Helpers")
local isEmpty = ____Helpers.isEmpty
____exports.Format = Format or ({})
____exports.Format.LEFT = 0
____exports.Format[____exports.Format.LEFT] = "LEFT"
____exports.Format.CENTER = 1
____exports.Format[____exports.Format.CENTER] = "CENTER"
____exports.Format.RIGHT = 2
____exports.Format[____exports.Format.RIGHT] = "RIGHT"
____exports.OutlineThickness = OutlineThickness or ({})
____exports.OutlineThickness.NONE = 0
____exports.OutlineThickness[____exports.OutlineThickness.NONE] = "NONE"
____exports.OutlineThickness.THIN = 1
____exports.OutlineThickness[____exports.OutlineThickness.THIN] = "THIN"
____exports.OutlineThickness.THICK = 2
____exports.OutlineThickness[____exports.OutlineThickness.THICK] = "THICK"
____exports.default = __TS__Class()
local FontWithPosition = ____exports.default
FontWithPosition.name = "FontWithPosition"
function FontWithPosition.prototype.____constructor(self, id, x, y, text, options)
    self.iconFormat = ____exports.Format.LEFT
    self.isDisabled = false
    self.color = {1, 1, 1, 1}
    self.id = id
    self.size = options and options.size
    self.filepath = options and options.filepath or TextManager:getDefaultFontFilepath()
    self.x = x
    self.y = y
    self.text = text
    self.format = options and options.format or ____exports.Format.LEFT
    self.icon = options and options.icon
    self.iconFormat = options and options.iconFormat or (self.format == ____exports.Format.CENTER and ____exports.Format.LEFT or self.format)
    local ____temp_12 = options and options.isDisabled
    if ____temp_12 == nil then
        ____temp_12 = false
    end
    self.isDisabled = ____temp_12
    self.outlineThickness = options and options.outlineThickness or ____exports.OutlineThickness.THIN
    if self.isDisabled then
        self:setDisabled(true)
    end
    self.color = options and options.color or ({1, 1, 1, 1})
    self.limit = options and options.limit
    self.alignMode = options and options.alignMode
end
function FontWithPosition.prototype.setDisabled(self, disabled)
    self.isDisabled = disabled
    self.color = disabled and ({0.5, 0.5, 0.5, 1}) or ({1, 1, 1, 1})
end
function FontWithPosition.prototype.printFont(self)
    local font = not isEmpty(self.size) and love.graphics.newFont(self.filepath, self.size) or love.graphics.newFont(self.filepath)
    love.graphics.setFont(font)
    local textW = font:getWidth(self.text)
    local textH = font:getHeight()
    local baseX = math.floor(self.x - self:getFormatOffset(textW))
    local baseY = math.floor(self.y - textH / 2)
    self:printOutline(baseX, baseY)
    love.graphics.setColor(self.color)
    local ____temp_21
    if not isEmpty(self.limit) then
        ____temp_21 = love.graphics.printf(
            self.text,
            baseX,
            baseY,
            self.limit,
            self.alignMode or "left"
        )
    else
        ____temp_21 = love.graphics.print(self.text, baseX, baseY)
    end
    self:renderIcon()
end
function FontWithPosition.prototype.printOutline(self, x, y)
    if self.outlineThickness == ____exports.OutlineThickness.NONE then
        return
    end
    love.graphics.setColor(0, 0, 0, self.color[4])
    local offsets = self.outlineThickness == ____exports.OutlineThickness.THICK and ({
        -2,
        -1,
        0,
        1,
        2
    }) or ({-1, 0, 1})
    for ____, ox in ipairs(offsets) do
        for ____, oy in ipairs(offsets) do
            do
                if ox == 0 and oy == 0 then
                    goto __continue9
                end
                love.graphics.print(self.text, x + ox, y + oy)
            end
            ::__continue9::
        end
    end
end
function FontWithPosition.prototype.getFormatOffset(self, textW)
    repeat
        local ____switch14 = self.format
        local iconWidth
        local ____cond14 = ____switch14 == ____exports.Format.LEFT
        if ____cond14 then
            return 0
        end
        ____cond14 = ____cond14 or ____switch14 == ____exports.Format.CENTER
        if ____cond14 then
            return textW / 2
        end
        ____cond14 = ____cond14 or ____switch14 == ____exports.Format.RIGHT
        if ____cond14 then
            iconWidth = self.iconFormat == ____exports.Format.RIGHT and not isEmpty(self.icon) and self.icon:getWidth() or 0
            return textW + iconWidth
        end
        do
            return 0
        end
    until true
end
function FontWithPosition.prototype.renderIcon(self)
    if isEmpty(self.icon) then
        return
    end
    love.graphics.setColor(self.color)
    repeat
        local ____switch17 = self.iconFormat
        local ____cond17 = ____switch17 == ____exports.Format.LEFT
        if ____cond17 then
            love.graphics.draw(
                self.icon,
                self.x - self.icon:getWidth() - 1,
                self.y - self.icon:getHeight() / 2
            )
            break
        end
        ____cond17 = ____cond17 or ____switch17 == ____exports.Format.RIGHT
        if ____cond17 then
            love.graphics.draw(
                self.icon,
                self.x - self.icon:getWidth() + 1,
                self.y - self.icon:getHeight() / 2
            )
            break
        end
    until true
end
return ____exports
